### DISCUSSION TITLE
[Bug]: T-Beam v1.2, M8N - GPS module initializes but doesn't work

### DISCUSSION THREAD

**Discussion by nikvostro** [meshtastic/firmware#3804] [Troubleshooting]
### Category

Other

### Hardware

T-Beam

### Firmware Version

2.3.7

### Description

T-Beam v1.2, M8N - GPS module initializes but doesn't work . GPS module **works fine** with https://github.com/Xinyuan-LilyGO/LilyGo-LoRa-Series/blob/master/firmware/NMEA_Reset firmware.

Troubleshooting steps performed:
1. Leaving for a long time >12 hours on the window. Several attempts - no luck
2. Factory reset, default settings, etc...
3. Reflash entire FW with pre-flashing "0" (blank) binary to overwrite flash contents
4. Lower ver 2.3.4, and 2.3.6 behaves the same way - no GPS positioning
5. [GPS_reset firmware ](https://github.com/Xinyuan-LilyGO/LilyGo-LoRa-Series/blob/master/firmware/NMEA_Reset/README.MD)feeds good NMEA flow with 20 satellites detected and 8 being used. 

**Important notice:** at the first start of freshly flashed meshtastic firmware I've obtained the device's GPS positing appear somewhere near the real. Like it was "cached" by the M8N module in time when it was driven by GPR_reset firmware a just couple of minutes before flashing Meshtastics firmware. Then GPS position was never reported.

From my POV the issue looks like related to a wrong communication with the module because it works fine (feeding NMEA via COM) with GPS reset firmware by Lilygo.

[DEBUG-20240505-144717.log](https://github.com/meshtastic/firmware/files/15212959/DEBUG-20240505-144717.log)

![2024-05-05 16_12_34-2024-05-05 14_27_04-VisualGPSView_](https://github.com/meshtastic/firmware/assets/83143316/71d582c1-8d6a-4bf9-b9ec-bbbb1f0a1124)

### Relevant log output

```Shell
Pease see attached log
```

**Comment by GPSFan** [comment]
I wish I could replicate your issue.
A few questions:
Is this the first time you have used this particular t-beam for Meshtastic, or any other firmware?
Roughly how long has this t-beam been powered up since you have had it? i.e. 1 hour, 12 hours, a day, a few days?
Are you using the tiny little GPS antenna that usually comes with t-beams?
How long did it take to get a fix with the LiliGo firmware?
Was the location of the antenna the same as with the Meshtastic firmware?
Your log shows a reboot at 14:49:47.601, the following reset_reason=reset implies you pushed the reset button, is this correct?

The LiliGo firmware and Meshtastic configure the M8N differently, so comparing operation of one vs the other is not straightforward.
Your Meshtastic log shows the M8N recognized (14:49:50.920 -> 14:49:54.246) and configured. That means that the t-beam is communicating correctly with the M8N.
If Meshtastic doesn't get an initial GPS lock within 15 minutes (900sec) it gives up and turns off the GPS, you can see this in your log where it sets "WANT GPS =0" at 15:04:49:837, followed by "Setting GPS power=0. 
If you have not had the t-beam powered "long enough" the backup battery will not be charged enough to keep the GPS data during a power off, or power save mode. It can take a long time to charge (controlled by the AXP2101). It will then be performing a "cold start" when turned on, if the signal levels are low (as shown by your GPS display) it can take a long time to acquire enough data to get a fix. Low signal levels are the result of a poor antenna (like the one shipped with the t-beams) and poor view of the sky (sitting in a window). If the GPS power is turned off, and the backup battery cannot sustain the BBR layer in the M8N, the configuration will revert to defaults. The config is saved to flash as well, but I have seen a corrupt BBR layer cause a reset to defaults rather than to Flash.
The LiliGo software resets the M8N's config to factory defaults and leaves the power on continuously, even with the tiny GPS antenna it will eventually get a fix.
This issue has come and gone in the firmware for a while, it seems to be related to the backup battery not being charged the first time the t-beam is used, as well as some unknowns that resolve themselves after a while.

**Comment by nikvostro** [comment]
Hello Ken!

Thank you for involving!

Let me answer by points:
1. _Is this the first time you have used this particular t-beam for Meshtastic, or any other firmware?_ - for T-Beam - first time, and after failing GPS I've just tried GPS_reset factory FW. No other experience and tests with other FWs.
2. _Roughly how long has this t-beam been powered up since you have had it? i.e. 1 hour, 12 hours, a day, a few days?_ - tried different timings up to several days powered on. I've also tried playing with **Position** settings section by making **Position broadcast** and **GPS update** intervals bigger (1800) - didn't help also.
3. _Are you using the tiny little GPS antenna that usually comes with t-beams?_ - yes. I'm waiting for a bigger one now to arrive. I thought it was the root cause but after flashing GPS_reset this became inactual.
4. _How long did it take to get a fix with the LiliGo firmware?_ - hmmm, didn't fixed that :(  Within 1-2 hours definetely.
5. _Was the location of the antenna the same as with the Meshtastic firmware?_ - yes
6. _Your log shows a reboot at 14:49:47.601, the following reset_reason=reset implies you pushed the reset button, is this correct?_ - can't remember exactly, but most probably yes. Or maybe I've changed some of device Settings that caused a reboot (?)

I thought that battery was related to RTC but thank you for clarifying! Just now I've measured it's coltage 2.4V before powering ON the T-Beam and since ~10 minutes it has grew up to 2.74V and continued raising very-very slowly. In specs I see the 3V of nominal voltage for this type of batt... Not sure if everything is fine here but why GPS work with GPS_reset FW...

What you could advise for the further investigation?

Attaching another log for almost 5 hrs online.
[DEBUG-20240504-171359.log](https://github.com/meshtastic/firmware/files/15217603/DEBUG-20240504-171359.log)

![20240506_094755](https://github.com/meshtastic/firmware/assets/83143316/b6c61242-b278-4cd1-8291-28c6cb94c58e)

**Comment by GPSFan** [comment]
"How long did it take to get a fix with the LiliGo firmware? - hmmm, didn't fixed that :( Within 1-2 hours definetely."

This is the key to your issue, Meshtastic will give up if no fix in 15 minutes. Because it took so long to get a fix with the LiliGo firmware, that indicates that your view of the sky is very poor. 

Try this, with either firmware take the t-beam outside with a good view of the sky, no trees, buildings nearby. Make sure the GPS antenna is pointed towards the sky. If it takes more than a minuet to get a fix, then either your antenna or M8N is faulty.

**Comment by nikvostro** [comment]
And how I can extend the time for getting GPS fix? I assume this linked with the some of broadcast intervals settings in **Position** section.
New GPS antenna is to come from day to day.

**Comment by GPSFan** [comment]
The use-cases for low signal level GPS, are few and far between. If your sky view in your use-case is always going to be poor, the reliability of the fix you get is also going to be poor. Meshtastic may not be the right tool for that use-case.
Get a good antenna and give it good view of the sky. Get a better GNSS receiver, the M8N is ok for an older generation receiver, but the M10, or one of the CN receivers from AliExpress can get a fix where the M8N will not. Of course if you are stuck with a t-beam you would have to unsolder the M8N and re-solder another module.

**Comment by nikvostro** [comment]
No, scenario is to use under the clear sky normally. But I'm a but surprised of a such poor performance of the factory antenna being used stationary and directed to the sky. Will see with the new antenna once it come.

Today sky isn't so clear and I can't get fix even with GPS_reset FW.
![2024-05-06 12_19_17-VisualGPSView](https://github.com/meshtastic/firmware/assets/83143316/4ab33428-dee7-41c1-badd-a6d01efc999c)

**Comment by nikvostro** [comment]
I've got new antenna! Much better and faster fix. GLONASS was inactive all the time with the factory antenna and now GLONASS satellites are even in use. Weather conditions: overcast, heavy clouds, no perciptations. Going to switch FW back to Meshtastic.

![2024-05-07 20_21_10-VisualGPSView](https://github.com/meshtastic/firmware/assets/83143316/c8be9f99-f8b2-42b2-b94a-1a9324056362)

**Comment by GPSFan** [comment]
Those signal levels are still really low,40-45 is usually considered adequate for a good fix.

**Comment by fifieldt** [comment]
@nikvostro , how is your GPS going?

**Reply by nikvostro** [reply]
Hi fifieldt! My GPS is fine, thanks! :)
New GPS antenna and extended time to get a fix resolved my issue. Usually 90-120secs is quite enough for my region when wheater is moreless fine.

**Reply by fifieldt** [reply]
Glad to hear it! Thanks for using Meshtastic :)

**Comment by GraphiteCA** [comment]
For what it's worth, I've upgraded both of my TBEAM's from 2.2.16 to 2.6.11, upon upgrading both, I'm nolonger able to get a GPS lock.

Two identical devices, that were up and running beside each other, with successful GPS locks, and then upon upgrading to 2.6.11, nolonger able to get a lock. No hardware changes (heck, no positional changes.)

I tried turning the GPS ON/OFF via triple-pressing the middle button, resetting the devices, moving them into a more favourable outdoor position with no appreciable change. 

While the devices could typically get a lock within a minute or two with the old firmware, the new firmware still are unable to get a lock after several hours.

@fifieldt - not sure if this is helpful for you, I'd be happy to do a side by side (or move one back down to 2.2.16 to verify?) if it'll help.

**Reply by fifieldt** [reply]
Thanks @GraphiteCA , most helpful would be logs - from first boot to 30+mins later. Plug in via USB go to https://flasher.meshtastic.org/ and click Open Serial monitor

**Reply by GraphiteCA** [reply]
> Thanks @GraphiteCA , most helpful would be logs - from first boot to 30+mins later. Plug in via USB go to https://flasher.meshtastic.org/ and click Open Serial monitor

Thanks for the reply @fifieldt and getting back so quickly-- per your request here's ~45min.
Sitting in the open, GPS antenna has a fairly unobstructed view of the sky.


[meshtastic-log-2025-08-26T23-01-53.929Z.log](https://github.com/user-attachments/files/21998023/meshtastic-log-2025-08-26T23-01-53.929Z.log)


Please don't hesitate to let me know if there's anything else I can do/provide to help! :)

**Reply by GraphiteCA** [reply]
Ofcourse, the law of demos strikes.... @fifieldt 
As soon as I post that one, the log updates that it eventually gets a lock at ~50 minute mark.
However, not only is it incredibly slow to lock, but the location has a huge margin of error and is centered approximately 3km off.

[meshtastic-log-2025-08-26T23-07-18.512Z.log](https://github.com/user-attachments/files/21998071/meshtastic-log-2025-08-26T23-07-18.512Z.log)

The previous version running (2.2.16) had a lock within a few minutes and was accurate within a few meters.

**Reply by GraphiteCA** [reply]
Diving into this a bit further, the lat= AND lon= coordinates in the log output are accurate.
I believe:
1. There's an issue with long lock time, I'm not sure the how/why.
2. The meshtastic IOS app does not appear to show the lat/lon correctly in the mesh map/position map/position log. It's ~2.5km off from the lat/lon from the logging on the device itself.

**Reply by GraphiteCA** [reply]
Ah Ok, I've resolved issue 2, it's a privacy setting on the channel in the meshtastic app itself. My apologies!

issue 1, however, is still strange, I'm firmly of the belief that there's a GPS regression as I'd previously mentioned. I don't believe it should take hours to achieve a GPS lock, where previous firmware versions were able to do it in minutes.

**Reply by fifieldt** [reply]
https://github.com/meshtastic/firmware/issues/7466 will help

**Reply by GraphiteCA** [reply]
Amazing, thank you. It looks like the issue is known, I eagerly await to test! :) Thanks again for your time!
